src/hooks.server.ts

```diff
@@ -66,3 +26,130 @@ export const handle: Handle = async ({ event, resolve }) => {
 
   return await resolve(event);
 };
+
+const handleJWTAuth: Handle = async ({ event, resolve }) => {
+  const cookieHeader = event.request.headers.get('cookie') ?? '';
+  const cookies = parse(cookieHeader);
+  const { activeRole: activeRoleCookie = null, user: userCookie } = cookies;
+
+  // try to get role with current JWT
+  if (userCookie) {
+    const user = await computeRolesFromCookies(userCookie, activeRoleCookie);
+    if (user) {
+      event.locals.user = user;
+      return await resolve(event);
+    }
+  }
+
+  // if we're already on the login page, don't redirect
+  // otherwise we get stuck in a redirect loop
+  return event.url.pathname.startsWith('/login') || event.url.pathname.startsWith('/auth')
+    ? await resolve(event)
+    : new Response(null, {
+        headers: {
+          location: `${base}/login`,
+        },
+        status: 307,
+      });
+};
+
+const handleSSOAuth: Handle = async ({ event, resolve }) => {
+  const cookieHeader = event.request.headers.get('cookie') ?? '';
+  const cookies = parse(cookieHeader);
+  const { activeRole: activeRoleCookie = null } = cookies;
+
+  // pass all cookies to the gateway, who can determine if we have any valid SSO tokens
+  const validationData = await reqGatewayForwardCookies<ReqValidateSSOResponse>(
+    '/auth/validateSSO',
+    cookieHeader,
+    event.request.url,
+  );
+
+  if (!validationData.success) {
+    console.log('Invalid SSO token, redirecting to SSO login UI page');
+    return new Response(null, {
+      headers: {
+        // redirectURL field from gateway response will contain our login UI URL
+        location: `${validationData.redirectURL}`,
+      },
+      status: 307,
+    });
+  }
+
+  // otherwise, we had a valid SSO token, so compute roles from returned JWT
+  // note, this sets a new JWT cookie each time
+  const user: BaseUser = {
+    id: validationData.userId ?? '',
+    token: validationData.token ?? '',
+  };
+
+  const roles = await computeRolesFromJWT(user, activeRoleCookie);
+
+  if (roles) {
+    console.log(`successfully SSO'd for user ${user.id}`);
+
+    event.locals.user = roles;
+
+    // create and set cookies
+    const userStr = JSON.stringify(user);
+    const userCookie = Buffer.from(userStr).toString('base64');
+    const cookieOpts: CookieSerializeOptions = {
+      httpOnly: false,
+      path: `${base}/`,
+      sameSite: 'none',
+    };
+    event.cookies.set('user', userCookie, cookieOpts);
+    // don't overwrite existing activeRole
+    if (!activeRoleCookie) {
+      event.cookies.set('activeRole', roles.defaultRole, cookieOpts);
+    }
+  } else {
+    event.locals.user = null;
+  }
+
+  return await resolve(event);
+};
+
+async function computeRolesFromCookies(
+  userCookie: string | null,
+  activeRoleCookie: string | null,
+): Promise<User | null> {
+  const userBuffer = Buffer.from(userCookie ?? '', 'base64');
+  const userStr = userBuffer.toString('utf-8');
+
+  try {
```

Comment:
I added this try catch so that the UI server wouldn't crash if a malformed `user` cookie got sent, i.e. `JSON.parse()` fails. The `computeRolesFromJWT` call is only within the try catch so it shares scope with `const baseUser`. 🤔

Reply:
I agree this is a little awkward to read, reads like it's trying to catch an exception from `computeRolesFromJWT`

Reply:
Ohhh. My bad. I didn't see the JSON.parse...